homeassistant/components/esphome/entry_data.py

```diff
@@ -125,18 +133,65 @@ async def async_update_static_infos(
         async_dispatcher_send(hass, signal, infos)
 
     @callback
-    def async_update_state(self, hass: HomeAssistant, state: EntityState) -> None:
+    def register_state_type(
+        self, state_type: type[EntityState], component_key: str
+    ) -> None:
+        """Register a state type."""
+        self.entity_state_type_to_component_key[state_type] = component_key
+        self.entity_state_type_name_to_component_key[
+            state_type.__name__
+        ] = component_key
+
+    @callback
+    def component_key_for_state(self, state: EntityState) -> str:
+        """Return the component key for a state."""
+        state_class_name = state.__class__.__name__
+
+        # We try to avoid doing a linear search for the
+        # the component key by looking up the class name
+        # in the mapping. If it's not there, we fall back
+        # to a linear search for the component key by doing
+        # an instance search over entity_state_type_to_component_key.
+        if state_class_name in self.entity_state_type_name_to_component_key:
+            return self.entity_state_type_name_to_component_key[state_class_name]
```

Comment:
Why not lookup by class?
```suggestion
        if component_key := self.entity_state_type_to_component_key.get(type(state)):
            return component_key
```

Reply:
Everything was already done by `component_key`

I just realized this might be a problem since buttons don't have their own state_type and use the generic `EntityState`

```
    await platform_async_setup_entry(
        hass,
        entry,
        async_add_entities,
        component_key="button",
        info_type=ButtonInfo,
        entity_type=EsphomeButton,
        state_type=EntityState,
    )
```

Reply:
It looks like Button the only one that uses that so I guess its not an issue

Reply:
I guess that makes sense since buttons are stateless